### DISCUSSION TITLE
How To Parse timestamp

### DISCUSSION THREAD

**Discussion by ffreality** [nanodbc/nanodbc#410] [Q&A] [answered]
`const int32 DataType = In_Result.column_datatype(Index_Column);` return `timestamp` as -2 from MSSQL.
I can't parse it with these because it gives `invalid datatype` error.
(I don't run them at the same time. I tried each of them one by one)

```
nanodbc::timestamp TimeStamp = In_Result.get<nanodbc::timestamp>(Index_Column);
nanodbc::date TimeStamp = In_Result.get<nanodbc::date>(Index_Column);
nanodbc::time TimeStamp = In_Result.get<nanodbc::time>(Index_Column);
```
On the other hand
```
nanodbc::string TimeStamp = In_Result.get<nanodbc::string>(Index_Column);
```
doesn't give error and `TimeStamp.lenght()` return `8`.
But if I try to use `c_str()`, string prints as white space.

Is there any suggestion about it ?

**Comment by mloskot** [comment]
@ffreality  Please, have a look at the technique currently covered in the tests:

https://github.com/nanodbc/nanodbc/blob/65d759295e3e876cf2207d7b38359c81ee49e4c1/test/mssql_test.cpp#L1110-L1128

https://github.com/nanodbc/nanodbc/blob/65d759295e3e876cf2207d7b38359c81ee49e4c1/test/postgresql_test.cpp#L357-L369

**Reply by ffreality** [reply]
> @ffreality Please, have a look at the technique currently covered in the tests:
> 
> https://github.com/nanodbc/nanodbc/blob/65d759295e3e876cf2207d7b38359c81ee49e4c1/test/mssql_test.cpp#L1110-L1128
> 
> https://github.com/nanodbc/nanodbc/blob/65d759295e3e876cf2207d7b38359c81ee49e4c1/test/postgresql_test.cpp#L357-L369

@mloskot
Thank you very much for your reply.

this sample code doesn't satisfy my question.
As you see, you created your column as `datetime` with name `d`
`create_table(connection, NANODBC_TEXT("test_datetime"), NANODBC_TEXT("d datetime"));`

If I have `datetime` type column, I can parse all values with `nanodbc::timestamp`. I am not asking that.

I am asking this for a column with `timestamp` variable type. Also you can not even `insert` a custom value to `timestamp`. You can only use `DEFAULT`.

`SQL Server Management Studio` shows `timestamp` as `0x0000000000009471`

**Reply by mloskot** [reply, marked as answer]
> I am asking this for a column with `timestamp` variable type. Also you can not even `insert` a custom value to `timestamp`. 

The `timestamp` type you use is not even a date and time data type available in SQL Server. The `timestamp` type you use is a very unfortunately named type and it is a mistake to think of it as of similar `timestamp` in PostgreSQL or other database systems.

From https://learn.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql
> The Transact-SQL [rowversion](https://learn.microsoft.com/en-us/sql/t-sql/data-types/rowversion-transact-sql?view=sql-server-ver16) data type is not a date or time data type. **timestamp** is a deprecated synonym for **rowversion**.

From https://learn.microsoft.com/en-us/sql/t-sql/functions/current-timestamp-transact-sql
> This function returns the current database system timestamp as a **datetime** value

Hence, we don't even cover SQL Server `timestamp` with tests and parsing it as `nanodbc::timestamp` value is a mistake.

**Reply by ffreality** [reply]
> > I am asking this for a column with `timestamp` variable type. Also you can not even `insert` a custom value to `timestamp`.
> 
> The `timestamp` type you use is not even a date and time data type available in SQL Server. The `timestamp` type you use is a very unfortunately named type and it is a mistake to think of it as of similar `timestamp` in PostgreSQL or other database systems.
> 
> From https://learn.microsoft.com/en-us/sql/t-sql/functions/date-and-time-data-types-and-functions-transact-sql
> 
> > The Transact-SQL [rowversion](https://learn.microsoft.com/en-us/sql/t-sql/data-types/rowversion-transact-sql?view=sql-server-ver16) data type is not a date or time data type. **timestamp** is a deprecated synonym for **rowversion**.
> 
> From https://learn.microsoft.com/en-us/sql/t-sql/functions/current-timestamp-transact-sql
> 
> > This function returns the current database system timestamp as a **datetime** value
> 
> Hence, we don't even cover SQL Server `timestamp` with tests and parsing it as `nanodbc::timestamp` value is a mistake.

@mloskot 
I looked at some documents and they say that, `timestamp` is actually `rowversion`. Developers use it to check if row changed since last retrive or not and it is not related with date or time. We can convert it with `CAST(stamp_column AS bigint)` C++ equivalent is long long int.

is there any possiblity to support it in the future ?

**Reply by mloskot** [reply]
@ffreality 
> I looked at some documents and they say that, timestamp is actually rowversion. 

Well, that is exactly what the documentation I quoted in https://github.com/nanodbc/nanodbc/discussions/410#discussioncomment-9974614 say, isn't it.

>We can convert it with CAST(stamp_column AS bigint) C++ equivalent is long long int.

Yes.

> is there any possiblity to support it in the future ?

Support what? I'm afraid, you need to make your feature request clearer as I don't understand what you are asking for.

Feel free to open new issue with your feature request, including clear description and, ideally, code SQL and C++ snippets explaining your desired use case.

**Reply by ffreality** [reply]
> @ffreality
> 
> > I looked at some documents and they say that, timestamp is actually rowversion.
> 
> Well, that is exactly what the documentation I quoted in [#410 (reply in thread)](https://github.com/nanodbc/nanodbc/discussions/410#discussioncomment-9974614) say, isn't it.
> 
> > We can convert it with CAST(stamp_column AS bigint) C++ equivalent is long long int.
> 
> Yes.
> 
> > is there any possiblity to support it in the future ?
> 
> Support what? I'm afraid, you need to make your feature request clearer as I don't understand what you are asking for.
> 
> Feel free to open new issue with your feature request, including clear description and, ideally, code SQL and C++ snippets explaining your desired use case.

Both of us speaking English and we talk about `timestamp` in MSSQL right now. I don't understand which section you couldn't understand. Currently, `nanodbc` doesn't support `timestamp` datatype. We are there.
I am asking a simple question. Do you guys have a plan to support `timestamp` in the future or should I stick with another library ?

**Reply by mloskot** [reply]
> Both of us speaking English

We do, but it does not guarantee that we will understand what we try to verbally communicate to each other, I'm afraid.

> and we talk about `timestamp` in MSSQL right now.

Yes, which is a [deprecated (alias) name](https://learn.microsoft.com/en-us/sql/t-sql/data-types/rowversion-transact-sql) for `rowversion` data type, which is a binary value of size of 8 bytes.

> Currently, nanodbc doesn't support `timestamp` datatype.

I dare to disagree based on what is feasible using the current `main` branch:

```cpp
// CREATE TABLE test_rowversion (myval int, myrv rowversion)
nanodbc::connection c(...);
execute(c, NANODBC_TEXT("INSERT INTO test_rowversion (myval) VALUES (123)"));
nanodbc::string sql = NANODBC_TEXT("SELECT myval, myrv FROM test_rowversion");
auto result = execute(c, sql);
while (result.next())
{
    auto v = result.get<int>(0);
    std::cout << std::dec << v << '\t';
    auto r = result.get<std::vector<std::uint8_t>>(1);
    std::cout << "0x";
    for (auto&& b : r) std::cout << std::hex << std::setfill('0') << std::setw(2) << (int)b;
    std::cout << std::endl;
}
```

![image](https://github.com/nanodbc/nanodbc/assets/80741/c8d8f828-777c-4399-b46d-a51cd9e64e89)

You're welcome!

> (...) or should I stick with another library ?

I'm in no position to tell users of an open source library that I happen to maintain what they should or should not do.

**Reply by ffreality** [reply]
yess, you are awasome !
I changed it to this for Unreal Engine.

```
std::vector<uint8_t> TempData = In_Result.get<std::vector<std::uint8_t>>(Index_Column);
						
std::stringstream StringStream;
for (auto&& EachByte : TempData)
{
	StringStream << std::hex << static_cast<int>(EachByte);
}

unsigned int TimeStampInt = std::stoul(StringStream.str(), nullptr, 16);
FString TimeStampString = UTF8_TO_TCHAR(StringStream.str().c_str());

EachData.Integer64 = TimeStampInt;
EachData.Preview = TimeStampString + " - " + FString::FromInt(TimeStampInt);
```

![Screenshot (79)](https://github.com/nanodbc/nanodbc/assets/10528447/7ffb210a-6ccf-4ab7-ab33-88b130cc31c6)

**Reply by mloskot** [reply]
Thank you!
